app/controllers/solidus_stripe/intents_controller.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+require 'stripe'
+
+class SolidusStripe::IntentsController < Spree::BaseController
+  include Spree::Core::ControllerHelpers::Order
+
+  before_action :load_payment_method
+
+  def payment_confirmation
+    intent = @payment_method.gateway.request { Stripe::PaymentIntent.retrieve(id: params[:payment_intent]) }
+    payment = @payment_method.find_in_progress_payment_for(current_order)
+
+    unless %w[confirm payment].include?(current_order.state.to_s)
+      redirect_to main_app.checkout_state_path(current_order.state)
+      return
+    end
+
+    current_order.state = :payment
+    # https://stripe.com/docs/payments/intents?intent=payment
+    case intent.status
+    when 'requires_payment_method'
+      # this state after an attempt at processing a payment means that it failed.
+      payment.failure! unless payment.failure?
+      ensure_state_is(payment, :processing)
```

Comment:
How is that we change the state to `:failure` but later expect it to be `:processing`?

Reply:
I guess it's related to:

> The most important difference is that on Stripe a failure is not a final state, rather just a way to start over.

However, that's something we're not handling yet, right?

Reply:
Good call! The expectation got it right, and the call to `failure!` is legacy of what I tried before understanding that caveat. But since that part is not covered by test yet it got through, fixing it 👍